### ISSUE TITLE
Repeatedly calling scipy.signal.resample uses more memory than expected

### ISSUE THREAD

**Issue by TimSC** [scipy/scipy#12297]
I have code that resamples many audio files using scipy.signal.resample. The code uses much more memory than I would expect to the point of freezing the OS. I've managed to simplify my code to a simple test case. This uses ~8Gb RAM after about 6 iterations but I would expect it should use much less memory than this.

I am using inputs of varying sizes, which seems to be significant. If I use an audio input of constant size there doesn't seem to be a problem.

```
from scipy.signal import resample
import numpy as np

def Test():

	shps = range(11473744, 1000, -2)

	for i, shp in enumerate(shps):
		signal = np.random.normal(size=(shp, 1))

		print (i, signal.shape)
		targetSamples = signal.shape[0]*44100//48000
		signal2 = resample(signal, targetSamples, axis=0)
		print (signal2.shape)
		del signal, signal2

if __name__=="__main__":
	Test()

```
I might have missed something obvious, so this might not even be a bug!

* Scipy 1.4.1 
* Numpy 1.18.4
* Python sys.version_info(major=3, minor=6, micro=9, releaselevel='final', serial=0)

**Comment by mreineck** [comment]
I'm not absolutely sure, but if the resampling process somehow calls `scipy.fft` internally, it might be caused by the module's internal plan caching. Currently `scipy.fft` caches plans for the last 16 different transforms it did, and for 1D transforms they are comparable in size to the transformed arrays (for "difficult" transform sizes even larger).

One way to test this hypothesis is to start your test with a smaller size and check whether the memory consumption still increases after 16 iterations.

**Comment by mreineck** [comment]
OK, inspecting the test with `perf top` shows that the FFT really is called internally. What I don't understand is why `resample` does the resampling in such a brute force fashion. I would have expected that this can be done chunk-wise using power-of-two chunk sizes ...
@peterbell10 do you happen to know more details?

**Comment by TimSC** [comment]
Dropping the start size 11473744 to a value 10 times smaller allows it to get to iteration 16. From then, it does seem to use a roughly constant amount of memory. 

Looking further, my bug is probably a dup of #5986

**Comment by mreineck** [comment]
> Perhaps limit the cache memory usage?

Perfectly reasonable suggestion. But coming up with a reasonable cache size is more difficult than it seems ... of course, we can make it flexible, but that requires an interface extension which users need to know about.

Personally I'd just drop the cache. It doesn't help that much, especially with the really expensive multi-D transforms.

For your concrete application I'm sure that there are ways to implement this within `scipy.signal` which are much more CPU- and memory-efficient than what's currently in there.

**Comment by mreineck** [comment]
One workaround for your situation (admittedly not pretty, but rather efficient, I'd expect) is to break the input into  chunks of 48.000 samples and resample them individually (the last one may be shorter of course).

**Comment by mreineck** [comment]
Something like this (completely untested, so take with a grain of salt):

```
from scipy.signal import resample
import numpy as np

def resample_chunks(signal):
    nchunks = (signal.shape[0] + 47999) // 48000
    res=[]
    for i in range(nchunks):
        beg = i*48000
        end = min((i+1)*48000, signal.shape[0])
        targetSamples = (end-beg)*44100 // 48000
        res.append(resample(signal[beg:end], targetSamples, axis=0))
    return np.concatenate(res, axis=0)

def Test():

	shps = range(11473744, 1000, -2)

	for i, shp in enumerate(shps):
		signal = np.random.normal(size=(shp, 1))

		print (i, signal.shape)
		#targetSamples = signal.shape[0]*44100//48000
		#signal2 = resample(signal, targetSamples, axis=0)
		signal2 = resample_chunks(signal)
		print (signal2.shape)
		del signal, signal2

if __name__=="__main__":
	Test()
```

**Comment by mreineck** [comment]
Naah, sorry, that's most likely not correct. Please ignore ...

**Comment by larsoner** [comment]
The resampling function (by using FFT) implicitly assumes periodicity, so resampling in chunks will not give the same result as resampling in one go, unfortunately. You might try using `resample_poly`, which uses a different algorithm (polyphase).

@peterbell10 any ideas about how to manage the cache? Maybe we do need to expose it or some functions publicly so that people can clear it.

**Comment by mreineck** [comment]
Even if we give the users the choice to manage the cache, we should assume by default that they don't know and/or don't care ... and the only safe default is "cache off" ... or perhaps "cache is limited to 1 MB". When I thought "well, 16 cached plans shouldn't hurt anyone", I clearly wasn't careful enough :-)

**Comment by mreineck** [comment]
> Yet for all the extra memory they take up, they actually don't contribute as much of a speedup. Only ~1-2% for 1D transforms in my testing.

Do you have an example script for this? This would be actually amazing :)
The convolution factors themselves are the result of a "long" FFT themselves, so if we do not cache them, we have to recalculate them, and this should account for (roughly) 33% of the whole transform. (A full Bluestein transform consists of 3 FFTs: one to compute the convolution factors, and two for the actual convolution of the data. The first one is independent of the data and can therefore be cached.)
The "chirp" twiddle factors may be another story, that's correct.

**Comment by mreineck** [comment]
I agree that it would be nice to directly cache `cfftp` and `rfftp` plans, because the Bluestein plans currently may hold redundant copies. Maybe there is a way of synthesizeing a `pocketfft_c` or `pocketfft_r` object from individually cached `cfftp` objects and convolution factors... but unless we are willing to give up the caching of the convolution factors, the memory consumption will not change fundamentally.

**Comment by peterbell10** [comment]
Actually, wait I was doing something really silly. I was generating input data with `np.arange(n*2).view(complex)` but since `arange` produces integers, I was getting entirely denormal numbers when viewed as floats so the FP math was way slower, but not for the chirp factors.

**Comment by peterbell10** [comment]
Indeed, it is about 1/3 speedup now that I test it again. In that case, caching the factors seperately is probably not a good idea since it means the bluestein plans won't be evicted from cache by regular sized plans, so will stick around for longer.
